pegen/grammar_visualizer.py

```diff
@@ -1,82 +1,114 @@
 import argparse
 import tokenize
+import functools
 import enum
 
 from pegen.tokenizer import Tokenizer
 from pegen.tokenizer import grammar_tokenizer
 from pegen.grammar2 import GrammarParser
 
-
-argparser = argparse.ArgumentParser(prog='pegen', description="Experimental PEG-like parser generator")
+argparser = argparse.ArgumentParser(prog='pegen', description="Pretty print the AST for a given PEG grammar")
 argparser.add_argument('filename', help="Grammar description")
 
 
 class Direction(enum.Enum):
-        UP = 0
-        DOWN = 1
-        NEUTRAL = 2
-        NONE = 3
-
-
-class GrammarPrinter:
-
-    def children(self, node):
+    UP = 0
+    DOWN = 1
+    NEUTRAL = 2
+    NONE = 3
+
+
+class ASTGrammarPrinter:
+    _direction_to_start_shape = {
+        Direction.UP: '┌',
+        Direction.DOWN: '└',
+        Direction.NEUTRAL: ' ',
+        Direction.NONE: '├'
+    }
+
+    @staticmethod
+    def children(node):
         for value in node:
             if isinstance(value, list):
                 yield from value
             else:
                 yield value
-    def nb_children(self, node):
-        return sum(self.nb_children(child) for child in self.children(node)) + 1
 
-    def name(self, node):
+    @staticmethod
+    def name(node):
+        # TODO: Make the grammar nodes be able to return this instead of hardcoding attributes
         if (name := getattr(node, "name", None)):
             return node.__class__.__name__ + f"({node.name})"
         elif (value := getattr(node, "value", None)):
             return node.__class__.__name__ + f"({node.value})"
         else:
             return node.__class__.__name__
 
+    @functools.lru_cache
+    def _total_number_of_children(self, node):
+        """Get the total number of children nodes under a given one recursively"""
+        return sum(self._total_number_of_children(child) for child in self.children(node)) + 1
 
-    def print_grammar(self, current_node, indent='', last=Direction.NEUTRAL, printer=print):
+    def _segment_children(self, node):
+        """Separate children of a given node in two groups so every group has roughly
+        the same number of nodes under it."""
 
-        child_sizes = {child: self.nb_children(child) for child in self.children(current_node)}
+        child_sizes = {child: self._total_number_of_children(child)
+                       for child in self.children(node)}
 
         up = sorted(child_sizes, key=lambda key: child_sizes[key])
```

Comment:
Why sort? I'd rather see the children displayed in the order in which they are stored in their parent. That order is significant in PEG grammars (since earlier alternatives must be tried before later ones).